rfcs/20190729-keras-preprocessing-redesign.md

```diff
@@ -0,0 +1,351 @@
+        target_mode:
+            - 'categorical' means that the inferred labels are
+                encoded as a categorical vector (e.g. for categorical_crossentropy).
+            - 'binary' means that the inferred labels (there can be only 2)
+                are encoded as binary scalars (e.g. for binary_crossentropy).
+        weight_column: Name of column with sample weight information.
+        color_mode: One of "grayscale", "rgb", "rgba". Default: "rgb".
+            Whether the images will be converted to
+            have 1, 3, or 4 channels.
+        batch_size: Size of the batches of data (default: 32).
+        image_size: Size to resize images to after they are read from disk.
+          Since the pipeline processes batches of images that must all have the same size,
+          this must be provided.
+        shuffle: Whether to shuffle the data (default: True)
+            If set to False, sorts the data in alphanumeric order.
+        seed: Optional random seed for shuffling and transformations.
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+
+def preview(self, data, save_to_directory=None, save_prefix=None, save_format='png'):
+    """Enables users to preview the image augmentation configuration.
+
+    # Arguments
+        data: Image data. Could be strings (a list of image paths), a list of PIL image instances,
+            a list of arrays, or a list of eager tensors.
+        save_to_directory: Directory to save transformed images. Mandatory if not in a notebook.
+            If in a notebook and this is not specified, images are displayed in-line.
+        save_prefix: String, filename prefix for saved images.
+        save_format: String, extension for saved images.
+    """
+```
+
+**Note:** `from_arrays` is not included since it is possible to transform Numpy data simply by calling the `ImagePipeline` object (like a layer).
+
+
+### Layers
+
+The new data augmentation layers will inherit `keras.layers.Layer` and work in a similar way.
+
+```python
+Resizing(height, width)  # Resize while distorting aspect ratio
+CenterCrop(height, width)  # Resize without distorting aspect ratio
+Rescaling(value)  # Divide by `value`
+RandomFlip(horizontal=False, vertical=False, seed=None)
+RandomTranslation(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomRotation(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomZoom(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomBrightness(amplitude=0., seed=None)
+RandomContrast(amplitude=0., seed=None)
+RandomSaturation(amplitude=0., seed=None)
+RandomWidth(amplitude=0., seed=None)  # Expand / shrink width while distorting aspect ratio
+RandomHeight(amplitude=0., seed=None)  # Expand / shrink height while distorting aspect ratio
```

Comment:
nit: amplitude is a weird parameter on some of these-- eg, what is the amplitude of a rotation, or a width resize? As these are separate layers whose params will diverge over time, does it make sense to use the "right" words here rather than biasing towards the same words?

Reply:
API consistency is important to reduce cognitive load and minimize surprises / looking up the docs. Is there a better universal word we could use in this context?

Note: this is also the reason why we use the keyword "kernel" throughout Keras even in places where it doesn't exactly apply.

Reply:
This is nice and easy to use, but I am a little concerned that there's no apparent way to specify these in absolute units: pixels, radians,... etc.

Exposing some way to skip the relative units would make it easier to build piplelines specified in absolute units without spending time questioning things like:
  - What the relative units are for each? How they're interpreted?
  - How does multi-scale training, or images with a different input range affect the pipeline?

> If larger than 1, it is rounded to one for the lower boundary (but not the higher boundary). 

For random zoom this comes out a little strange. If I want a uniform scale in `[1/2, 2]` I can set `amplitude=[1/2, 1]`. But,  IIUC, the random part here is linear so 1/3 of images are shrunk, and 2/3 are expanded. A log-scale option for random zoom would be nice to have.